### PR TITLE
Undo/redo

### PR DESCRIPTION
**Pull request by bkmartinjr**
Implement undo/redo user functions.   Changes:
* Refactor reuducers, using a new "cascading" meta reducer.  
    * The `controls` reducer is now split into: universe, world, categoricalSelectionState, crossfilter, colors and controls (misc UI state)
    * net effect is that the factored reducers (eg, universe) are now in `state.universe` rather than `state.controls.universe`
* Add undo/redo support, which includes:
    * the new undoable meta reducer which manages the state history
    * configuration of which actions side-effect the history
    * Rework a few actions and action generators to improve history UI - ie, allow history to change appropriately based upon user action
    * add Redo/Undo buttons to the graph component

Initial cut at UI:
* Reset will clear the history
* All user actions should be undoable/redoable

There remain a number of edge cases & known issues:
* brushable histogram does not correctly clear its brush state when reset (via either Reset or Undo).   #627 
* lasso selector does not correctly update its paint state in Undo or Redo (the selection is correctly handled -- this is only an issue with the lasso layer painting).  #663

### CODE REVIEW

File: client/src/components/brushableHistogram/index.js [hunk 1]
```diff
@@ -15,13 +15,13 @@ import actions from "../../actions";
...
+  world: state.world,
   scatterplotXXaccessor: state.controls.scatterplotXXaccessor,
   scatterplotYYaccessor: state.controls.scatterplotYYaccessor,
-  crossfilter: state.controls.crossfilter,
+  crossfilter: state.crossfilter,
   differential: state.differential,
-  colorAccessor: state.controls.colorAccessor,
-  obsAnnotations: _.get(state.controls.world, "obsAnnotations", null)
+  colorAccessor: state.colors.colorAccessor,
+  obsAnnotations: _.get(state.world, "obsAnnotations", null)
```
Comment (colinmegill) [inline client/src/components/brushableHistogram/index.js]: `state.world` is really nice.

File: client/src/components/brushableHistogram/index.js [hunk 2]
```diff
@@ -289,7 +290,8 @@ class HistogramBrush extends React.Component {
       .call(
         d3
           .brushX()
-          .on("brush", this.onBrush(field, x.invert).bind(this))
+          .on("start", this.onBrush(field, x.invert, "start").bind(this))
+          .on("brush", this.onBrush(field, x.invert, "brush").bind(this))
```
Comment (colinmegill) [inline client/src/components/brushableHistogram/index.js]: Requesting a comment here re: why all 3 are necessary

File: client/src/components/categorical/categorical.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ import * as globals from "../../globals";
 import Category from "./category";
 
 @connect(state => ({
-  categoricalSelectionState: state.controls.categoricalSelectionState
+  categoricalSelectionState: state.categoricalSelectionState
```
Comment (colinmegill) [inline client/src/components/categorical/categorical.js]: A point for later, but `categoricalSelection` is probably the right name here, state is redundant.

File: client/src/reducers/categoricalSelectionState.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    case "categorical metadata filter all of these": {
+      const newCategoricalSelectionState = {
+        ...state,
+        [action.metadataField]: {
+          ...state[action.metadataField],
+          categorySelected: Array.from(
+            state[action.metadataField].categorySelected
+          ).fill(true)
```
Comment (colinmegill) [inline client/src/reducers/categoricalSelectionState.js]: Suggest a comment on this line for posterity

File: client/src/reducers/crossfilter.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            "scalar",
+            world.varData.col(gene).asArray(),
+            Float32Array
+          ),
+        state
+      );
+      return crossfilter;
+    }
+
+    case "clear differential expression": {
```
Comment (colinmegill) [inline client/src/reducers/crossfilter.js]: Do we think this belongs in the crossfilter reducer?

File: client/src/components/brushableHistogram/index.js (hunk 2 above)
Comment (bkmartinjr) [reply client/src/components/brushableHistogram/index.js]: will add.    TL;DR - needed for History.

The essence is that we want to know when the drag starts so we can push the Redux state onto the history stack.   We don't want to push new history on each brush/drag, but only on the first one.

File: client/src/components/categorical/categorical.js (hunk 3 above)
Comment (bkmartinjr) [reply client/src/components/categorical/categorical.js]: I'll fix, but for the record, I used the name that was previously in the code (ie, I didn't change it).

Comment (colinmegill) [reply client/src/components/categorical/categorical.js]: Yup, this is my name

File: client/src/reducers/categoricalSelectionState.js (hunk 4 above)
Comment (bkmartinjr) [reply client/src/reducers/categoricalSelectionState.js]: I'm not sure what you are proposing - can you elaborate?

File: client/src/reducers/crossfilter.js (hunk 5 above)
Comment (bkmartinjr) [reply client/src/reducers/crossfilter.js]: yes, it does.    All code that changes the crossfilter state is in this reducer.

Note that we also handle this action in other reducers, to change other state.

File: client/src/reducers/categoricalSelectionState.js (hunk 4 above)
Comment (colinmegill) [reply client/src/reducers/categoricalSelectionState.js]: This is one of the 'chewier' lines, those reading from a non JS background might appreciate a plaintext reading of this operation. If you disagree feel free to skip.

File: client/src/components/categorical/categorical.js (hunk 3 above)
Comment (bkmartinjr) [reply client/src/components/categorical/categorical.js]: fixed.

File: client/src/reducers/categoricalSelectionState.js (hunk 4 above)
Comment (bkmartinjr) [reply client/src/reducers/categoricalSelectionState.js]: lol, you wrote this code :-)

I'll add comments.

Comment (bkmartinjr) [reply client/src/reducers/categoricalSelectionState.js]: OK, added a comment about the intent (to all four places we do something similar).  LMK if that resolves it, or propose a comment if you want something different (I wasn't sure if "chewie" referred to the intent or the code).

File: client/src/reducers/crossfilter.js (hunk 5 above)
Comment (colinmegill) [reply client/src/reducers/crossfilter.js]: Right, yes, that's great.

File: client/src/reducers/categoricalSelectionState.js (hunk 4 above)
Comment (colinmegill) [reply client/src/reducers/categoricalSelectionState.js]: Those are perfect!

**Pull request merged**